flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/runtimefilter/RuntimeFilterCodeGenerator.scala

```diff
@@ -0,0 +1,126 @@
+import org.apache.flink.table.runtime.operators.CodeGenOperatorFactory
+import org.apache.flink.table.runtime.util.StreamRecordCollector
+import org.apache.flink.table.types.logical.RowType
+
+/** Operator code generator for runtime filter operator. */
+object RuntimeFilterCodeGenerator {
+  def gen(
+      ctx: CodeGeneratorContext,
+      buildType: RowType,
+      probeType: RowType,
+      probeIndices: Array[Int]): CodeGenOperatorFactory[RowData] = {
+    val probeGenProj = ProjectionCodeGenerator.generateProjection(
+      ctx,
+      "RuntimeFilterProjection",
+      probeType,
+      RowTypeUtils.projectRowType(probeType, probeIndices),
+      probeIndices)
+    ctx.addReusableInnerClass(probeGenProj.getClassName, probeGenProj.getCode)
+
+    val probeProjection = newName("probeToBinaryRow")
+    ctx.addReusableMember(s"private transient ${probeGenProj.getClassName} $probeProjection;")
+    val probeProjRefs = ctx.addReusableObject(probeGenProj.getReferences, "probeProjRefs", null)
+    ctx.addReusableOpenStatement(
+      s"$probeProjection = new ${probeGenProj.getClassName}($probeProjRefs);")
+
+    val buildEnd = newName("buildEnd")
+    ctx.addReusableMember(s"private transient boolean $buildEnd;")
+    ctx.addReusableOpenStatement(s"$buildEnd = false;")
+
+    val filter = newName("filter")
+    val filterClass = classOf[BloomFilter].getCanonicalName
+    ctx.addReusableMember(s"private transient $filterClass $filter;")
+
+    val collector = newName("collector")
+    val collectorClass = classOf[StreamRecordCollector[_]].getCanonicalName
+    ctx.addReusableMember(s"private transient $collectorClass<$ROW_DATA> $collector;")
+    ctx.addReusableOpenStatement(s"$collector = new $collectorClass<>(output);")
+
+    val input1Item = "input1"
+    val input2Item = "input2"
+
+    val processElement1Code =
+      s"""
+         |if ($buildEnd) {
+         |    throw new IllegalStateException("Should not build ended.");
+         |}
+         |if ($filter == null && !$input1Item.isNullAt(1)) {
+         |    $filter = $filterClass.fromBytes($input1Item.getBinary(1));
+         |}
+         |""".stripMargin
+
+    val processElement2Code =
+      s"""
+         |if (!$buildEnd) {
+         |    throw new IllegalStateException("Should build ended.");
+         |}
+         |if ($filter != null) {
+         |    final int hashCode = $probeProjection.apply($input2Item).hashCode();
+         |    if ($filter.testHash(hashCode)) {
+         |        $collector.collect($input2Item);
+         |    }
+         |} else {
+         |    $collector.collect($input2Item);
+         |}
+         |""".stripMargin
+
+    val nextSelectionCode = s"return $buildEnd ? $INPUT_SELECTION.SECOND : $INPUT_SELECTION.FIRST;"
+
+    val endInputCode1 =
+      s"""
+         |if ($buildEnd) {
+         |    throw new IllegalStateException("Should not build ended.");
+         |}
+         |LOG.info("Finish build phase.");
```

Comment:
BloomFilter build completed.

Reply:
Maybe using `RuntimeFilter` is more appropriate here.